### PR TITLE
Tier2 - CEPH-83575332 - CephFS-Mirroring-HA

### PR DESCRIPTION
**Pull request by hkadam134**
Logs:  http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-2W53C5/
Polarion ID: CEPH-83575332 - Create a 2 Node - CephFS Mirroring HA from Cephadm shell and check if the Service has come up as expected and validate mirroring is established to a remote cluster and all the information regarding sync status

# Description

Please include Automation development guidelines. Source of Test case - New Feature/Regression Test/Close loop of customer BZs
<details>

<summary>click to expand checklist</summary>

- [ ] Create a test case in Polarion reviewed and approved.
- [ ] Create a design/automation approach doc. Optional for tests with similar tests already automated.
- [ ] Review the automation design
- [ ] Implement the test script and perform test runs
- [ ] Submit PR for code review and approve
- [ ] Update Polarion Test with Automation script details and update automation fields
- [ ] If automation is part of Close loop, update BZ flag qe-test_coverage “+” and link Polarion test
</details>

### CODE REVIEW

**Comment by hkadam134** [comment]
Latest Logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-SG3EKD/
Reduced the usage of redundant setup functions and introduced "deploy_cephfs_mirroring" and "destroy_cephfs_mirroring" functions.

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 1]
```diff
@@ -478,6 +486,102 @@ def validate_synchronization(
...
+            f"fs mirror status {fs_name}@{filesystem_id} -f json",
+        )
+        data = json.loads(out)
+        log.info(data)
+        dir_count = data.get("snap_dirs", {}).get("dir_count")
+        if not data.get("snap_dirs"):
+            raise CommandFailed("Unable to get the Snap Dir")
+        if dir_count:
+            log.info(f"Snap Directory Count : {dir_count}")
+            log.info("Synchronization is successful")
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:533]: this message seems not appropriate as in this module we are just getting snap dir count, we are not comparing it with some source to confirm sync was successful.

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 2]
```diff
@@ -478,6 +486,102 @@ def validate_synchronization(
...
+        Returns:
+        - int: The count of synced directories
+
+        Raises:
+        - CommandFailed: If the directory count is not found or synchronization fails
+        """
+        fsid = self.get_fsid(cephfs_mirror_node)
+        filesystem_id = self.get_filesystem_id_by_name(source_clients, fs_name)
+        asok_file = self.get_asok_file(cephfs_mirror_node, fsid, daemon_name)
+        log.info("Validate the synced dir from cephfs-mirror status")
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:515]: This message had to be "Get the synced dir count..." instead

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 3]
```diff
@@ -478,6 +486,102 @@ def validate_synchronization(
...
+        data = json.loads(out)
+        log.info(data)
+        dir_count = data.get("snap_dirs", {}).get("dir_count")
+        if not data.get("snap_dirs"):
+            raise CommandFailed("Unable to get the Snap Dir")
+        if dir_count:
+            log.info(f"Snap Directory Count : {dir_count}")
+            log.info("Synchronization is successful")
+        else:
+            log.error("Snap Directory Count is not matching, Synchronization failed")
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:535]: As we are not matching dir_count with some input param, the error message could be, "Failed to get dir_count from mirror status, synchronization failed" ?

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 4]
```diff
@@ -478,6 +486,102 @@ def validate_synchronization(
...
+            raise CommandFailed(
+                "Snap Directory Count is not matching, Synchronization failed"
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:537]: same as above

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 5]
```diff
@@ -478,6 +486,102 @@ def validate_synchronization(
...
+            f"{fs_name}@{filesystem_id} {peer_uuid} -f json"
+        )
+        out, _ = cephfs_mirror_node.exec_command(sudo=True, cmd=cmd)
+        data = json.loads(out)
+        for path, status in data.items():
+            last_synced_snap = status.get("last_synced_snap")
+            if last_synced_snap:
+                return last_synced_snap.get("name")
+
+        log.error("Snapshot not found or not synced")
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:582]: can we elaborate snapshot as "last synced snapshot not found or not synced"?

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py [hunk 6]
```diff
@@ -478,6 +486,102 @@ def validate_synchronization(
...
+        raise CommandFailed("Snapshot not found or not synced")
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:583]: same as above

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+                cephfs_mirror_node[0],
+                source_fs,
+                fsid,
+                asok_file1,
+                filesystem_id,
+                peer_uuid,
+            )
+        )
+        log.info(
+            f"Total Synced directory found on {cephfs_mirror_node[0].node.hostname} is"
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py:228]: As its not dir count, but snapshot names we printing, can the info start as "Synced directory found on ..."

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+                cephfs_mirror_node[1],
+                source_fs,
+                fsid,
+                asok_file2,
+                filesystem_id,
+                peer_uuid,
+            )
+        )
+        log.info(
+            f"Total Synced directory found on {cephfs_mirror_node[1].node.hostname} is "
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py:244]: same as above

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+            results.append(result)
+            log.info(f"Status of {snapshot_name}: {result}")
+
+        if all(successes):
+            log.info(
+                "Test Completed Successfully. All snapshots and data are synced to the target cluster"
+            )
+        else:
+            log.info(
+                "Test Failed. Snapshots or Data are missing on the target cluster."
```
Comment (sumabai) [inline tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py:291]: We had to error here as log.error ?

**Comment by hkadam134** [comment]
Entire cephfs mirroring ha suite Logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-5OQJ9K/

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py (hunk 1 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:533]: Removed this "            log.info("Synchronization is successful")"

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py (hunk 4 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:537]: Updated.

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py (hunk 5 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:582]: fixed

File: tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py (hunk 6 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/cephfs_mirroring_utils.py:583]: fixed

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py (hunk 7 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py:222]: fixed

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py (hunk 8 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py:238]: fixed

File: tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py (hunk 9 above)
Comment (hkadam134) [reply tests/cephfs/cephfs_mirroring/test_cephfs_mirroring_configure_cephfs_mirroring_ha.py:285]: fixed

**Comment by hkadam134** [comment]
> Do we require changes in other test scripts as we have modified Utils

Yes Amar - Made the required changes in all the scripts under the suite.

**Comment by hkadam134** [comment]
Ran the Entire Suite after fixing the required changes. - http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-LTL6G2/

1 Failure can be ignored as it's a product issue and we have a BZ filed

**Review by AmarnatReddy — APPROVED** [review]

**Pull request merged**